### PR TITLE
Do not block onchain order parsing on unknown events

### PR DESCRIPTION
**Pull request by fedgiac**
### Context

We pick up orders from the eth-flow contract by interpreting it as an instance of a generic wrapper contract `CoWSwapOnchainOrders`. This generic contracts defines events that should be interpreted as the creation of valid orders, directly from the blockchain.
However, the eth-flow contract can also emit other events that are not part of the `CoWSwapOnchainOrders` interface (namely order refunds). In the current implementation, our `CoWSwapOnchainOrders` contract instance (representing the contract at the eth-flow address) tries to interpret these events as an event that belongs to `CoWSwapOnchainOrders` itself and, when failing, returns an error that stalls the process of picking up events from the blockchain.

### This PR

We try to solve this issue by ignoring related errors on demand. This behavior must be opted in explicitly by the event handler instance and only applies to errors that are triggered by a failed ABI decoding of the error.  

### Requirements before merging

This PR requires changes to ethcontracts as otherwise we would copy over a lot of code from there. The reason for this is that one of the needed functions that decodes contract events squashes all errors together and returns `Result<Vec<Event>,_>` rather than the needed `Result<Vec<Event,_>>`.

I don't have access to the ethcontracts repo so I can't push a PR, but if given access I'll open one and release a new version once accepted. The changes would look like [this](https://github.com/fedgiac/ethcontract-rs/commit/17c9c8e61b840701c3ce0e90aefe4e33f726dceb) (but I expect to make small changes).

### Test Plan

A new e2e test is introduced to simulate the bad behavior.

### CODE REVIEW

**Review by MartinquaXD — COMMENTED** [review]
How strict is solidity with enforcing correctness of events? For example if it's impossible to emit an event that has the expected type but is encoded incorrectly then we could assume that all events that fail to decode are events which are irrelevant to us, right?
Then we could always ignore decode errors.

**Comment by fedgiac** [comment]
> How strict is solidity with enforcing correctness of events? For example if it's impossible to emit an event that has the expected type but is encoded incorrectly then we could assume that all events that fail to decode are events which are irrelevant to us, right?

A well-behaving smart contract would never create an event of a specific type that fails decoding. The only way to do that would be to emit a malformed event manually in an assembly block, and I've yet to see some projects doing it.
So you are proposing to not give any choice and default to `OnEventDecodingError::Ignore`? I don't see many drawbacks to that, the only reasonable one is that we might fail to notice that we are indexing the wrong contract.

**Comment by MartinquaXD** [comment]
> So you are proposing to not give any choice and default to OnEventDecodingError::Ignore?

Yes, that seems reasonable to me. AFAIK we can't do anything to fix the decoding errors and we can't do anything useful with the malformed events so I think ignoring those errors is acceptable.
@nlordell does that seem alright to you too?

File: crates/shared/src/event_handling.rs [hunk 1]
```diff
@@ -402,6 +432,21 @@ where
     }
 }
 
+/// Filter function that always returns true unless set to ignore event decoding
+/// errors. In this latter case, all errors are preserved except those that come
+/// from decoding an invalid event for the contract.
+fn filter_decoding_errors<E>(
```
[the remark this answers is not in the document]
Comment (fedgiac) [reply crates/shared/src/event_handling.rs]: Updated! Feel free to comment again if you don't like the changes.

**Comment by nlordell** [comment]
Hmm, if possible I would prefer an alternative approach to this solution. Specifically, in the `EventStream`, we really do want to error out on ABI decoding errors (I would say that the raw log stream would be the "general" stream to use if you wanted to handle decoding errors).

That being said, I think there is an easier approach to solve this issue that doesn't involve `ethcontract-rs` changes.

Specifically, you can define a new event wrapper:
```rust
pub enum PermissiveEthFlowEvent {
    Event(ethflow::Event),
    Other(RawLog),
}

impl ParseLog for PermissiveEthFlowEvent {
    fn parse_log(log: RawLog) -> Result<Self, ExecutionError> {
        match ethflow::Event::parse_log(log.clone()) {
            Ok(event) => Ok(Self::Event(event)),
            Err(_) => Ok(Self::Other(log)),
        }
    }
}
```

Then, you would create an event handler around this `PermissiveEthFlowEvent` instead of the usual `AllEvents` flow. Filtering would be done in the code that handles these events.

**Comment by nlordell** [comment]
Here is an example of a custom `ParseLog` implementation that we already use for Uniswap V3 liquidity - so this pattern can be replicated here as well:

https://github.com/cowprotocol/services/blob/019a1d13d632b72f999a175852089a5be026bcd5/crates/shared/src/sources/uniswap_v3/event_fetching.rs#L28

**Comment by fedgiac** [comment]
Closed in favor of  #943.

**Pull request closed without merging**